File: midi-script/Namespaces/TakeLane.py
from __future__ import absolute_import

from .Clip import Clip
from .Interface import Interface


class TakeLane(Interface):
    @staticmethod
    def serialize_take_lane(lane):
        if lane is None:
            return None

        return {
            "id": Interface.save_obj(lane),
            "name": lane.name,
        }

    def __init__(self, c_instance, socket):
        super(TakeLane, self).__init__(c_instance, socket)

    def get_arrangement_clips(self, ns):
        return [Clip.serialize_clip(clip) for clip in ns.arrangement_clips]

    def create_audio_clip(self, ns, file_path, start_time):
        clip = ns.create_audio_clip(file_path, start_time)
        return Clip.serialize_clip(clip)

    def create_midi_clip(self, ns, start_time, length):
        clip = ns.create_midi_clip(start_time, length)
        return Clip.serialize_clip(clip)

File: src/ns/take-lane.ts
import { Ableton } from "../index.js";
import { Namespace } from "./index.js";
import { Clip, RawClip } from "./clip.js";

export interface GettableProperties {
  arrangement_clips: RawClip[];
  name: string;
}

export interface TransformedProperties {
  arrangement_clips: Clip[];
}

export interface SettableProperties {
  name: string;
}

export interface ObservableProperties {
  arrangement_clips: RawClip[];
  name: string;
}

export interface RawTakeLane {
  readonly id: string;
  readonly name: string;
}

export class TakeLane extends Namespace<
  GettableProperties,
  TransformedProperties,
  SettableProperties,
  ObservableProperties
> {
  constructor(
    ableton: Ableton,
    public readonly raw: RawTakeLane,
  ) {
    super(ableton, "take-lane", raw.id);

    this.transformers = {
      arrangement_clips: (clips) => clips.map((c) => new Clip(ableton, c)),
    };

    this.cachedProps = {
      arrangement_clips: true,
    };
  }

  /** Creates an audio clip in this take lane's arrangement at `startTime`. */
  public async createAudioClip(filePath: string, startTime: number) {
    const rawClip = await this.sendCommand("create_audio_clip", {
      file_path: filePath,
      start_time: startTime,
    });
    return new Clip(this.ableton, rawClip);
  }

  /** Creates an empty MIDI clip in this take lane's arrangement. */
  public async createMidiClip(startTime: number, length: number) {
    const rawClip = await this.sendCommand("create_midi_clip", {
      start_time: startTime,
      length: length,
    });
    return new Clip(this.ableton, rawClip);
  }
}
